llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,466 @@
+//===----- AEIWaWRegRewriter.cpp - Rewrite Register to not include multiple Defs
+//---------===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of a Loop Register rewriting pass, that reduces WAW physical
+// register conflicts, so that software pipelining compact loops better.
+//
+//===----------------------------------------------------------------------===//
+
+#include "AIEBaseInstrInfo.h"
+#include "AIEBaseRegisterInfo.h"
+
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/CodeGen/LiveDebugVariables.h"
+#include "llvm/CodeGen/LiveIntervals.h"
+#include "llvm/CodeGen/LiveRegMatrix.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstr.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineLoopInfo.h"
+#include "llvm/CodeGen/MachineOperand.h"
+#include "llvm/CodeGen/Passes.h"
+#include "llvm/CodeGen/SlotIndexes.h"
+#include "llvm/CodeGen/TargetInstrInfo.h"
+#include "llvm/CodeGen/TargetSubtargetInfo.h"
+#include "llvm/MC/MCRegister.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include <llvm/CodeGen/MachineBasicBlock.h>
+#include <map>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "aie-wawreg-rewrite"
+
+namespace {
+
+cl::opt<bool> DisablePass(
```

Comment:
I think it is not clear if your pass is in the optimization pipeline or not. My suggestion is to add this pass using our `AIE2PassConfig`. Also, you can move this flag to `AIE2TargetMachine.cpp` to control the insertion of the pass in the optimization pipeline. You can also use something like:

```
cl::opt<bool> EnableWAWRegRewrite(
    "aie-wawreg-rewrite",
    cl::desc("Enable the WAW Register Renaming in loops (default true)"),
    cl::init(true), cl::Hidden);
```

Using the pass manager, we can use (assuming a proper place for insertion):
```
  if (EnableWAWRegRewrite)
    addPass(createAIEWawRegRewriter());
```

Then we can disable using `-aie-wawreg-rewrite=false`.

Reply:
made the pass registration clear

Reply:
I moved it `AIE2TargetMachine.cpp` for better readability